src/lib.rs

```diff
@@ -292,179 +292,259 @@ mod tests {
-/// preserving const-ness, and casting between the correct types.
-/// Implementing this is required in order to use `try_ref_from_ptr!` or
-/// `try_mut_from_ptr!`.
-pub(crate) trait CastPtr {
-    type RustType;
+/// Used to mark that pointer to a [`Castable`]'s underlying `Castable::RustType` is provided
+/// to C code as a pointer to a `Box<Castable::RustType>`.
+pub(crate) struct BoxCastTypeMarker;
 
-    fn cast_mut_ptr(ptr: *mut Self) -> *mut Self::RustType {
-        ptr as *mut _
-    }
-}
+/// Used to mark that a pointer to a [`Castable`]'s underlying `Castable::RustType` is provided
+/// to C code as a pointer to an `Arc<Castable::RustType>`.
+pub(crate) struct ArcCastTypeMarker;
 
-/// CastConstPtr represents a subset of CastPtr, for when we can only treat
-/// something as a const (for instance when dealing with Arc).
-pub(crate) trait CastConstPtr {
-    type RustType;
+/// A trait for marking the type of a pointer to a [`Castable`]'s underlying `Castable::RustType`
+/// that is provided to C code, either a [`BoxCastTypeMarker`] when it is a pointer to a `Box<_>`,
+/// or an [`ArcCastTypeMarker`] when it is a pointer to an `Arc<_>`.
+trait CastTypeMarker {}
 
-    fn cast_const_ptr(ptr: *const Self) -> *const Self::RustType {
-        ptr as *const _
-    }
+impl CastTypeMarker for BoxCastTypeMarker {}
+
+impl CastTypeMarker for ArcCastTypeMarker {}
+
+/// `Castable` represents the relationship between a snake case type (like [`client::rustls_client_config`])
+/// and the corresponding Rust type (like [`rustls::ClientConfig`]), specified as the associated type
+/// `RustType`. Each `Castable` also has an associated type `CastType` using one of the
+/// [`CastTypeMarker`] types, [`BoxCastTypeMarker`] or [`ArcCastTypeMarker`].
+///
+/// An implementation of `Castable` that uses [`BoxCastTypeMarker`] indicates that when we give C code
+/// a pointer to the relevant `RustType` `T`, that it is actually a `Box<T>`. Similarly, an
+/// implementation of `Castable` that uses [`ArcCastTypeMarker`] means that when we give C code a
+/// pointer to the relevant type, that it is actually an `Arc<T>`.
+///
+/// By using an associated type on `Castable` to communicate this we can use the type system to
+/// guarantee that a single type can't implement `Castable` for more than one [`CastTypeMarker`],
+/// since this would be a conflicting trait implementation and rejected by the compiler.
+///
+/// This trait allows us to avoid using `as` in most places, and ensures that when we cast, we're
+/// preserving const-ness, and casting between the correct types. Implementing this is required in
+/// order to use `try_ref_from_ptr!` or `try_mut_from_ptr!` and several other helpful cast-related
+/// conversion helpers.
+pub(crate) trait Castable {
+    /// Indicates whether to use `Box` or `Arc` when giving a pointer to C code for the underlying
+    /// `RustType`.
+    type CastType: CastTypeMarker;
```

Comment:
"CastType" might not be the best name for this idea and the associated marker types but it's the one I picked for a first pass.

It feels a bit like this is more of a wrapper type for the cast, so perhaps the names should have "wrapper" somewhere?

Reply:
Yeah I definitely got turned around by the `CastType` name a couple times playing with this code.

How about `Ownership`? With possible values OwnershipBox, OwnershipArc, and OwnershipRef.

Or `PointerStyle`?

We could also potentially change `RustType` to `Pointee` to go with `PointerStyle`.

Reply:
> How about Ownership? With possible values OwnershipBox, OwnershipArc, and OwnershipRef.

I like those :+1:

How do you feel about the top-level `Castable` name? Should we lay out some paint swatches for that bike shed while we're at it? :sweat_smile:

Reply:
>> How do you feel about the top-level Castable name? 
>
> I like Castable a lot!

Great :man_artist: - Resolving this thread since I pushed a revision with the new `Ownership` names.